lib/kcerts/ssh.go

```diff
@@ -0,0 +1,197 @@
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type sshCache struct {
+	PID       int       `json:"pid"`
+	Sock      string    `json:"sock"`
+	TimeStamp time.Time `json:"time_stamp"`
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, ttl time.Duration) (string, int, error) {
+	// TODO(adam): sniff out dead agents and sneak pids
+	envSSHSock := os.Getenv("SSH_AUTH_SOCK")
+	envSSHPID := os.Getenv("SSH_AGENT_PID")
+	if envSSHSock != "" && envSSHPID != "" {
+		pid, err := strconv.Atoi(envSSHPID)
+		if err != nil {
+			return "", 0, fmt.Errorf("%s is not a valid pid: %w", envSSHPID, err)
+		}
+		return envSSHSock, pid, nil
+	}
+
+	sshEnkitCache, isFresh, err := store.Get(SSHCacheKey)
+	if err != nil {
+		return "", 0, fmt.Errorf("error fetching cache: %w", err)
+	}
+
+	if isFresh {
+		f, err := os.OpenFile(filepath.Join(sshEnkitCache, "ssh.json"), os.O_RDWR, 0750)
+		if err != nil && !os.IsNotExist(err) {
+			return "", 0, fmt.Errorf("error opening cache: %w", err)
+		}
+		if err == nil {
+			cacheContent, err := ioutil.ReadAll(f)
+			if err != nil {
+				return "", 0, fmt.Errorf("error reading cache: %w", err)
+			}
+			var sshCache sshCache
+			if err := json.Unmarshal(cacheContent, &sshCache); err != nil {
```

Comment:
rather than `os.OpenFile`, `ioutil.ReadAll` followed by `json.Unmarshal`, I recommend you use: `marshal.UnmarshalFile(filepath.Join(sshEnkitCache, "ssh.json"), &sshCache)`, from `import "github.com/enfabrica/enkit/lib/config/marshal"`.